solidity/contracts/bridge/WalletCoordinator.sol

```diff
@@ -0,0 +1,542 @@
+    struct DepositSweepProposal {
+        bytes20 walletPubKeyHash;
+        DepositKey[] depositsKeys;
+    }
+
+    /// @notice Helper structure representing a plain-text deposit key.
+    ///         Each deposit can be identified by their 32-byte funding
+    ///         transaction hash (Bitcoin internal byte order) an the funding
+    ///         output index (0-based).
+    /// @dev Do not confuse this structure with the deposit key used within the
+    ///      Bridge contract to store deposits. Here we have the plain-text
+    ///      components of the key while the Bridge uses a uint representation of
+    ///      keccak256(fundingTxHash | fundingOutputIndex) for gas efficiency.
+    struct DepositKey {
+        bytes32 fundingTxHash;
+        uint32 fundingOutputIndex;
+    }
+
+    /// @notice Helper structure holding deposit extra data required during
+    ///         deposit sweep proposal validation. Basically, this structure
+    ///         is a combination of BitcoinTx.Info and relevant parts of
+    ///         Deposit.DepositRevealInfo.
+    struct DepositExtra {
+        BitcoinTx.Info fundingTx;
+        bytes8 blindingFactor;
+        bytes20 walletPubKeyHash;
+        bytes20 refundPubKeyHash;
+        bytes4 refundLocktime;
+    }
+
+    /// @notice Mapping that holds addresses allowed to submit proposals.
+    mapping(address => bool) public isProposalSubmitter;
+
+    /// @notice Mapping that holds wallet time locks. The key is a 20-byte
+    ///         wallet public key hash. The value is a UNIX timestamp defining
+    ///         the moment until which the wallet is locked and cannot receive
+    ///         new proposals. The value of 0 means the wallet is not locked
+    ///         and can receive a proposal at any time.
+    mapping(bytes20 => uint32) public walletLock;
+
+    /// @notice Handle to the Bridge contract.
+    Bridge public bridge;
+
+    /// @notice Handle to the WalletRegistry contract.
+    EcdsaWalletRegistry public walletRegistry;
+
+    /// @notice Determines the deposit sweep proposal validity time. In other
+    ///         words, this is the worst-case time for a deposit sweep during
+    ///         which the wallet is busy and cannot take another actions. This
+    ///         is also the duration of the time lock applied to the wallet
+    ///         once a new deposit sweep proposal is submitted.
+    ///
+    ///         For example, if a deposit sweep proposal was submitted at
+    ///         2 pm and depositSweepProposalValidity is 4 hours, the next
+    ///         proposal (of any type) can be submitted after 6 pm.
+    uint32 public depositSweepProposalValidity;
+
+    /// @notice The minimum time that must elapse since the deposit reveal
+    ///         before a deposit becomes eligible for a deposit sweep.
+    ///
+    ///         For example, if a deposit was revealed at 9 am and depositMinAge
+    ///         is 2 hours, the deposit is eligible for sweep after 11 am.
```

Comment:
> Does it mean the off-chain client does not have to validate confirmations of inputs

Nope, this check is not about that. The confirmations have to be checked anyway if we want to avoid producing signatures of invalid BTC transactions.

> Or is it just an additional check to ensure the wallet can't get blocked with data that will not pass the off-chain validation anyway.

This check ensures the block finality for the happy path. If we wait 2 hours before taking a proper deposit for the sweep, we are sure at least 20 ETH blocks and 6 BTC blocks have been mined in the meantime. In such a case, the BTC confirmations will be checked off-chain anyway though that will be just formality.

Reply:
Just to clarify: this is about Ethereum block finality and Bitcoin block finality is just the bonus, correct?

If that's the case, let's add a `@dev` section explaining it, please.

Reply:
See 0848de33